decisions/0003-files-in-JSON-output.md

```diff
@@ -0,0 +1,112 @@
+# 3. Files in JSON Output
+
+Date: 2019-08-27
+
+## Status
+
+🤔 Proposed
+
+## Context
+
+When users submit forms created by Form Builder the data from the user needs to
+be handed over to other systems for processing. This is fulfilled by the JSON
+output generated. One category of data which can be present is that of files in
+the form of documents, images and other files users may upload as part of their
+form submission.
+
+In order to hand over these files to other systems they need to be somehow
+present in the JSON.
+
+Considerations taken on board were security, ease of implementing and ease of
+integration. We need to consider that the data is sent securely and should only
+be accessible by the intended recipient. The solution should be cost effective
+and therefore cheap to implement and maintain. Integrating systems should able
+to receive files with ease therefore able to provide cost effective solutions.
+
+We have discussed the pros and cons of 4 possible options:
+
+1. Include files in JSON document as Base64 encoded strings
+
+Pros:
+
+- Super simple with a single POST of entire payload
+- We can delete all data once submission has been handed over
+
+Cons:
+
+- Potentially a large POST body
+- Files will be Base64 encoded which will increase size of payload and
+processing
+- Clients will have to decode file
+
+2. Embed signed many use AWS S3 URL in JSON
+
+Pros:
+
+- Lightweight JSON payload
+- S3 deals with files
+
+Cons:
+
+- We must retain files for client to pick up
```

Comment:
We should be explicit here that we'd be storing unencrypted user data for the first time in Form Builder. Not having to do this is a pro for options 1 and 4 as well.

Or we could encrypt the files in S3 with the same pre-shared secret that we're using to encrypt the JSON payload? That rather assumes that the application receiving the JSON is also fetching the files, or else encourages people to share that key more widely. It would work in our first use case (assuming the optics API supports file upload?), and it's a more secure default position to start from, so maybe we should take this approach from the start.

Reply:
As the files are encrypted, option 2 is no longer possible I will add an option 5 with this point present